File: command_gate/command_gate_node.py
#!/usr/bin/env python3
"""
command_gate_node.py — Python implementation of the CommandGate safety relay.

Gate logic: gate_open = (heartbeat_ok OR !require_heartbeat) AND (enabled OR !require_enable)

NOTE: Assumes single-threaded executor. Loading into a MultiThreadedExecutor would require
mutex protection on heartbeat_ok_, enabled_, and channel last_fallback_bytes.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

import rclpy
from rclpy.node import Node
from rosidl_runtime_py.utilities import get_message
from std_msgs.msg import Bool
from std_srvs.srv import SetBool


@dataclass
class Channel:
    input_topic: str
    output_topic: str
    message_type: str
    fallback_topic: str
    msg_class: object = None
    input_sub: object = None
    output_pub: object = None
    fallback_sub: object = None
    last_fallback_msg: object = None
    zero_msg: object = None
    warned_drop: bool = False


class CommandGateNode(Node):
    def __init__(self):
        super().__init__('command_gate')
        self._declare_and_get_params()
        self._setup_channels()
        self._setup_heartbeat()
        self._setup_enable()

        timer_period = self._heartbeat_timeout / 4.0
        self._heartbeat_timer = self.create_timer(timer_period, self._heartbeat_check_cb)

        self.get_logger().info(
            f'CommandGateNode started: {len(self._channels)} channel(s), '
            f'require_heartbeat={self._require_heartbeat}, '
            f'require_enable={self._require_enable}'
        )

    # ------------------------------------------------------------------ params

    def _declare_and_get_params(self) -> None:
        self.declare_parameter('num_channels', 1)
        self.declare_parameter('require_heartbeat', True)
        self.declare_parameter('heartbeat_timeout', 1.0)
        self.declare_parameter('heartbeat_topic', '~/heartbeat')
        self.declare_parameter('require_enable', False)
        self.declare_parameter('enable_topic', '~/enable')

        self._num_channels: int = self.get_parameter('num_channels').value
        self._require_heartbeat: bool = self.get_parameter('require_heartbeat').value
        self._heartbeat_timeout: float = self.get_parameter('heartbeat_timeout').value
        self._heartbeat_topic: str = self.get_parameter('heartbeat_topic').value
        self._require_enable: bool = self.get_parameter('require_enable').value
        self._enable_topic: str = self.get_parameter('enable_topic').value

        # Gate state
        self._enabled: bool = False
        self._heartbeat_ok: bool = False
        # Initialize as timed-out so gate starts closed when require_heartbeat=True
        self._last_heartbeat_time = (
            self.get_clock().now().nanoseconds * 1e-9
            - self._heartbeat_timeout - 1.0
        )

    # ---------------------------------------------------------------- channels

    def _setup_channels(self) -> None:
        self._channels: list[Channel] = []

        for i in range(self._num_channels):
            prefix = f'channel_{i}'
            self.declare_parameter(f'{prefix}.input_topic', '')
            self.declare_parameter(f'{prefix}.output_topic', '')
            self.declare_parameter(f'{prefix}.message_type', '')
            self.declare_parameter(f'{prefix}.fallback_topic', '')

            ch = Channel(
                input_topic=self.get_parameter(f'{prefix}.input_topic').value,
                output_topic=self.get_parameter(f'{prefix}.output_topic').value,
                message_type=self.get_parameter(f'{prefix}.message_type').value,
                fallback_topic=self.get_parameter(f'{prefix}.fallback_topic').value,
            )

            if not ch.input_topic or not ch.output_topic or not ch.message_type:
                self.get_logger().warn(
                    f'Channel {i} missing required parameters, skipping.'
                )
                continue

            try:
                ch.msg_class = get_message(ch.message_type)
            except Exception as exc:
                self.get_logger().warn(
                    f'Channel {i}: cannot resolve message type '
                    f'"{ch.message_type}": {exc}, skipping.'
                )
                continue

            ch.output_pub = self.create_publisher(ch.msg_class, ch.output_topic, 10)

            ch.input_sub = self.create_subscription(
                ch.msg_class, ch.input_topic,
                lambda msg, c=ch: self._on_input(msg, c), 10
            )

            if ch.fallback_topic:
                ch.fallback_sub = self.create_subscription(
                    ch.msg_class, ch.fallback_topic,
                    lambda msg, c=ch: self._on_fallback(msg, c), 1
                )

            try:
                ch.zero_msg = ch.msg_class()
            except Exception as exc:
                self.get_logger().warn(
                    f'Channel {i}: cannot build zero message for '
                    f'"{ch.message_type}": {exc}'
                )

            self._channels.append(ch)

    # -------------------------------------------------------- heartbeat / enable

    def _setup_heartbeat(self) -> None:
        from std_msgs.msg import Empty  # local import keeps top-level clean
        self._heartbeat_sub = self.create_subscription(
            Empty, self._heartbeat_topic, self._on_heartbeat_cb, 10
        )

    def _setup_enable(self) -> None:
        self._enable_sub = self.create_subscription(
            Bool, self._enable_topic, self._on_enable_topic_cb, 10
        )
        self._set_enabled_srv = self.create_service(
            SetBool, '~/set_enabled', self._on_set_enabled_cb
        )

    # ---------------------------------------------------------------- callbacks

    def _on_heartbeat_cb(self, _msg) -> None:
        was_open = self._is_gate_open()
        self._last_heartbeat_time = self.get_clock().now().nanoseconds * 1e-9
        self._heartbeat_ok = True
        now_open = self._is_gate_open()
        if not was_open and now_open:
            self._on_gate_state_change(False, True, self._open_reason('heartbeat received'))

    def _heartbeat_check_cb(self) -> None:
        was_open = self._is_gate_open()
        if self._require_heartbeat:
            now_sec = self.get_clock().now().nanoseconds * 1e-9
            elapsed = now_sec - self._last_heartbeat_time
            prev_ok = self._heartbeat_ok
            self._heartbeat_ok = elapsed < self._heartbeat_timeout
            now_open = self._is_gate_open()
            if was_open and not now_open:
                self._on_gate_state_change(True, False, 'heartbeat timeout')
            elif not prev_ok and self._heartbeat_ok and not was_open and now_open:
                self._on_gate_state_change(False, True, self._open_reason('heartbeat received'))

    def _on_enable_topic_cb(self, msg: Bool) -> None:
        was_open = self._is_gate_open()
        self._enabled = msg.data
        now_open = self._is_gate_open()
        if was_open != now_open:
            reason = self._open_reason('enabled via topic') if now_open else 'disabled via topic'
            self._on_gate_state_change(was_open, now_open, reason)

    def _on_set_enabled_cb(
        self, request: SetBool.Request, response: SetBool.Response
    ) -> SetBool.Response:
        was_open = self._is_gate_open()
        self._enabled = request.data
        now_open = self._is_gate_open()
        if was_open != now_open:
            reason = self._open_reason('enabled via service') if now_open else 'disabled via service'
            self._on_gate_state_change(was_open, now_open, reason)
        response.success = True
        response.message = f'enabled set to {self._enabled}'
        return response

    def _on_input(self, msg, ch: Channel) -> None:
        if self._is_gate_open():
            ch.output_pub.publish(msg)

    def _on_fallback(self, msg, ch: Channel) -> None:
        ch.last_fallback_msg = msg

    # ---------------------------------------------------------------- gate logic

    def _is_gate_open(self) -> bool:
        hb_ok = self._heartbeat_ok or not self._require_heartbeat
        en_ok = self._enabled or not self._require_enable
        return hb_ok and en_ok

    def _open_reason(self, trigger: str) -> str:
        if self._require_heartbeat and self._require_enable:
            if 'heartbeat' in trigger:
                return 'heartbeat received and already enabled'
            return f'heartbeat alive and {trigger}'
        return trigger

    def _on_gate_state_change(self, was_open: bool, now_open: bool, reason: str) -> None:
        if now_open:
            self.get_logger().info(f'Gate OPEN — {reason}')
            for ch in self._channels:
                ch.warned_drop = False
        else:
            self.get_logger().info(f'Gate CLOSED — {reason}')
            for ch in self._channels:
                self._publish_fallback_or_zero(ch)

    def _publish_fallback_or_zero(self, ch: Channel) -> None:
        if ch.last_fallback_msg is not None:
            ch.output_pub.publish(ch.last_fallback_msg)
        elif ch.zero_msg is not None:
            ch.output_pub.publish(ch.zero_msg)
        elif not ch.warned_drop:
            self.get_logger().warn(
                f'Gate closed on channel [{ch.output_topic}]: '
                f'no fallback or zero message available, dropping.'
            )
            ch.warned_drop = True


def main(args=None):
    rclpy.init(args=args)
    node = CommandGateNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/command_gate_node.cpp
#include "command_gate/command_gate_node.hpp"

#include <chrono>
#include <string>

#include "rclcpp/serialization.hpp"
#include "ackermann_msgs/msg/ackermann_drive.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"
#include "rclcpp_components/register_node_macro.hpp"
#include "std_msgs/msg/float64.hpp"

namespace command_gate {

CommandGateNode::CommandGateNode(const rclcpp::NodeOptions & options)
: Node("command_gate", options)
{
  declareAndGetParams();
  setupChannels();
  setupHeartbeat();
  setupEnable();

  heartbeat_timer_ = create_wall_timer(
    std::chrono::duration<double>(heartbeat_timeout_ / 4.0),
    [this]() { heartbeatCheckTimerCb(); });

  RCLCPP_INFO(
    get_logger(),
    "CommandGateNode started: %zu channel(s), require_heartbeat=%s, require_enable=%s",
    channels_.size(),
    require_heartbeat_ ? "true" : "false",
    require_enable_ ? "true" : "false");
}

void CommandGateNode::declareAndGetParams()
{
  declare_parameter("num_channels", 1);
  declare_parameter("require_heartbeat", true);
  declare_parameter("heartbeat_timeout", 1.0);
  declare_parameter("heartbeat_topic", std::string("~/heartbeat"));
  declare_parameter("require_enable", false);
  declare_parameter("enable_topic", std::string("~/enable"));

  require_heartbeat_ = get_parameter("require_heartbeat").as_bool();
  heartbeat_timeout_ = get_parameter("heartbeat_timeout").as_double();
  require_enable_ = get_parameter("require_enable").as_bool();

  enabled_ = false;
  heartbeat_ok_ = false;
  // Initialize as timed-out so gate starts closed when require_heartbeat=true
  last_heartbeat_time_ =
    get_clock()->now() - rclcpp::Duration::from_seconds(heartbeat_timeout_ + 1.0);
}

void CommandGateNode::setupChannels()
{
  int num_channels = get_parameter("num_channels").as_int();

  for (int i = 0; i < num_channels; ++i) {
    std::string prefix = "channel_" + std::to_string(i);
    declare_parameter(prefix + ".input_topic", std::string(""));
    declare_parameter(prefix + ".output_topic", std::string(""));
    declare_parameter(prefix + ".message_type", std::string(""));
    declare_parameter(prefix + ".fallback_topic", std::string(""));

    Channel ch;
    ch.input_topic = get_parameter(prefix + ".input_topic").as_string();
    ch.output_topic = get_parameter(prefix + ".output_topic").as_string();
    ch.message_type = get_parameter(prefix + ".message_type").as_string();
    ch.fallback_topic = get_parameter(prefix + ".fallback_topic").as_string();

    if (ch.input_topic.empty() || ch.output_topic.empty() || ch.message_type.empty()) {
      RCLCPP_WARN(get_logger(), "Channel %d missing required parameters, skipping.", i);
      continue;
    }

    ch.output_pub = create_generic_publisher(ch.output_topic, ch.message_type, rclcpp::QoS(10));

    // idx is captured by value; lambdas index into channels_ after construction is done
    const int idx = static_cast<int>(channels_.size());

    ch.input_sub = create_generic_subscription(
      ch.input_topic, ch.message_type, rclcpp::QoS(10),
      [this, idx](std::shared_ptr<rclcpp::SerializedMessage> msg) {
        if (isGateOpen()) {
          channels_[idx].output_pub->publish(*msg);
        }
      });

    if (!ch.fallback_topic.empty()) {
      ch.fallback_sub = create_generic_subscription(
        ch.fallback_topic, ch.message_type, rclcpp::QoS(1),
        [this, idx](std::shared_ptr<rclcpp::SerializedMessage> msg) {
          // Deep copy — original shared_ptr is reclaimed after the callback returns
          channels_[idx].last_fallback_msg =
            std::make_shared<rclcpp::SerializedMessage>(*msg);
        });
    }

    ch.zero_msg = makeZeroMessage(ch.message_type);
    channels_.push_back(std::move(ch));
  }
}

void CommandGateNode::setupHeartbeat()
{
  std::string heartbeat_topic = get_parameter("heartbeat_topic").as_string();
  heartbeat_sub_ = create_subscription<std_msgs::msg::Empty>(
    heartbeat_topic, rclcpp::QoS(10),
    [this](const std_msgs::msg::Empty::SharedPtr msg) { onHeartbeatCb(msg); });
}

void CommandGateNode::setupEnable()
{
  std::string enable_topic = get_parameter("enable_topic").as_string();
  enable_sub_ = create_subscription<std_msgs::msg::Bool>(
    enable_topic, rclcpp::QoS(10),
    [this](const std_msgs::msg::Bool::SharedPtr msg) { onEnableTopicCb(msg); });

  set_enabled_srv_ = create_service<std_srvs::srv::SetBool>(
    "~/set_enabled",
    [this](
      const std_srvs::srv::SetBool::Request::SharedPtr request,
      std_srvs::srv::SetBool::Response::SharedPtr response) {
      onSetEnabledCb(request, response);
    });
}

std::shared_ptr<rclcpp::SerializedMessage>
CommandGateNode::makeZeroMessage(const std::string & type)
{
  auto result = std::make_shared<rclcpp::SerializedMessage>();
#define TRY(MsgT, str) \
  if (type == (str)) { \
    rclcpp::Serialization<MsgT> s; MsgT m; s.serialize_message(&m, result.get()); \
    return result; \
  }
  TRY(geometry_msgs::msg::Twist,                  "geometry_msgs/msg/Twist")
  TRY(geometry_msgs::msg::TwistStamped,           "geometry_msgs/msg/TwistStamped")
  TRY(ackermann_msgs::msg::AckermannDrive,        "ackermann_msgs/msg/AckermannDrive")
  TRY(ackermann_msgs::msg::AckermannDriveStamped, "ackermann_msgs/msg/AckermannDriveStamped")
  TRY(std_msgs::msg::Float64,                     "std_msgs/msg/Float64")
#undef TRY
  RCLCPP_WARN(
    get_logger(), "Type '%s' not in known list — zero-init unavailable", type.c_str());
  return nullptr;
}

bool CommandGateNode::isGateOpen() const
{
  bool hb_ok = heartbeat_ok_ || !require_heartbeat_;
  bool en_ok = enabled_ || !require_enable_;
  return hb_ok && en_ok;
}

std::string CommandGateNode::openReason(const std::string & trigger) const
{
  if (require_heartbeat_ && require_enable_) {
    if (trigger.find("heartbeat") != std::string::npos) {
      return "heartbeat received and already enabled";
    }
    return "heartbeat alive and " + trigger;
  }
  return trigger;
}

void CommandGateNode::onGateStateChange(
  [[maybe_unused]] bool was_open, bool now_open, const std::string & reason)
{
  if (now_open) {
    RCLCPP_INFO(get_logger(), "Gate OPEN — %s", reason.c_str());
    for (auto & ch : channels_) {
      ch.warned_drop = false;
    }
  } else {
    RCLCPP_INFO(get_logger(), "Gate CLOSED — %s", reason.c_str());
    for (auto & ch : channels_) {
      publishFallbackOrZero(ch);
    }
  }
}

void CommandGateNode::publishFallbackOrZero(Channel & ch)
{
  if (ch.last_fallback_msg) {
    ch.output_pub->publish(*ch.last_fallback_msg);
  } else if (ch.zero_msg) {
    ch.output_pub->publish(*ch.zero_msg);
  } else if (!ch.warned_drop) {
    RCLCPP_WARN(
      get_logger(),
      "Gate closed on channel [%s]: no fallback or zero message available, dropping.",
      ch.output_topic.c_str());
    ch.warned_drop = true;
  }
}

void CommandGateNode::onHeartbeatCb(const std_msgs::msg::Empty::SharedPtr /*msg*/)
{
  bool was_open = isGateOpen();
  last_heartbeat_time_ = get_clock()->now();
  heartbeat_ok_ = true;
  bool now_open = isGateOpen();
  if (!was_open && now_open) {
    onGateStateChange(false, true, openReason("heartbeat received"));
  }
}

void CommandGateNode::heartbeatCheckTimerCb()
{
  bool was_open = isGateOpen();
  if (require_heartbeat_) {
    double elapsed = (get_clock()->now() - last_heartbeat_time_).seconds();
    bool prev_ok = heartbeat_ok_;
    heartbeat_ok_ = (elapsed < heartbeat_timeout_);
    bool now_open = isGateOpen();
    if (was_open && !now_open) {
      onGateStateChange(true, false, "heartbeat timeout");
    } else if (!prev_ok && heartbeat_ok_ && !was_open && now_open) {
      onGateStateChange(false, true, openReason("heartbeat received"));
    }
  }
}

void CommandGateNode::onEnableTopicCb(const std_msgs::msg::Bool::SharedPtr msg)
{
  bool was_open = isGateOpen();
  enabled_ = msg->data;
  bool now_open = isGateOpen();
  if (was_open != now_open) {
    std::string reason = now_open ? openReason("enabled via topic") : "disabled via topic";
    onGateStateChange(was_open, now_open, reason);
  }
}

void CommandGateNode::onSetEnabledCb(
  const std_srvs::srv::SetBool::Request::SharedPtr request,
  std_srvs::srv::SetBool::Response::SharedPtr response)
{
  bool was_open = isGateOpen();
  enabled_ = request->data;
  bool now_open = isGateOpen();
  if (was_open != now_open) {
    std::string reason = now_open ? openReason("enabled via service") : "disabled via service";
    onGateStateChange(was_open, now_open, reason);
  }
  response->success = true;
  response->message = std::string("enabled set to ") + (enabled_ ? "true" : "false");
}

}  // namespace command_gate

RCLCPP_COMPONENTS_REGISTER_NODE(command_gate::CommandGateNode)
